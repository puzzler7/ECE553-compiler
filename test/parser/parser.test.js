const fs = require('fs');
const { spawnSync } = require('child_process');
const { test, it } = require('@jest/globals');

const getRelativePathOfTests = (dir) => {
    return fs.readdirSync(dir)
             .filter(fileName => /^.*\.tig$/.exec(fileName))
             .map(fileName => dir + fileName);

}

const negativeTestPattern = /\/\*.*error.*\*\//;
const binaryName = `parserParse-${process.platform}`;

describe('When building with Mlton', () => {
    it('checks that Mlton exists and builds lexer without crashing', () => {
        const checkMltonExists = spawnSync('mlton')
        expect(checkMltonExists.error).toBeFalsy();
        expect(checkMltonExists.stdout.toString().includes('MLton')).toBeTruthy();

        const checkBuild = spawnSync('mlton', [
            '-default-ann',
            'allowVectorExps true',
            '-output',
            `build/${binaryName}`,
            'src/parser/sources.mlb'
        ]);
        expect(checkBuild.error).toBeFalsy();
        expect(checkBuild.stderr.toString().length).toBe(0);
    });
})

describe('The textbook tests', () => {
    const cases = getRelativePathOfTests('test/testcases/');
    /*
    const negativeTests = cases.filter(testFile => 
        negativeTestPattern.test(fs.readFileSync(testFile, {encoding: 'utf8'}))
    )
    const positiveTests = cases.filter()
    */

    test.each(cases)(
        'parses %s correctly',
        (testFile) => {
            
            const parseResults = spawnSync(`build/${binaryName}`, [testFile])
            expect(parseResults.error).toBeFalsy();
            expect(parseResults.stderr.toString()).toBe('');
            expect(parseResults.stdout.toString()).toMatchSnapshot();
            expect(errorMessagePattern.test(parseResults.stdout.toString())).toBeFalsy();
        }
    );
})