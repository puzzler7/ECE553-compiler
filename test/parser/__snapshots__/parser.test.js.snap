// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The textbook tests parses test/testcases/queens.tig correctly 1`] = `
"test/testcases/queens.tig:16.16:syntax error: inserting  SEMICOLON
test/testcases/queens.tig:16.26:syntax error: replacing  EQ with  ASSIGN
test/testcases/queens.tig:17.12:syntax error: inserting  SEMICOLON
test/testcases/queens.tig:18.15:syntax error: inserting  SEMICOLON
test/testcases/queens.tig:23.21:syntax error: inserting  PLUS
test/testcases/queens.tig:25.17:syntax error: replacing  EQ with  ASSIGN
test/testcases/queens.tig:25.32:syntax error: replacing  EQ with  ASSIGN
test/testcases/queens.tig:25.49:syntax error: replacing  EQ with  ASSIGN
test/testcases/queens.tig:28.22:syntax error: inserting  SEMICOLON
test/testcases/queens.tig:32.8:syntax error: inserting  SEMICOLON
"
`;

exports[`The textbook tests parses test/testcases/test1.tig correctly 1`] = `""`;

exports[`The textbook tests parses test/testcases/test2.tig correctly 1`] = `""`;

exports[`The textbook tests parses test/testcases/test3.tig correctly 1`] = `""`;

exports[`The textbook tests parses test/testcases/test4.tig correctly 1`] = `
"test/testcases/test4.tig:8.20:syntax error: inserting  PLUS
test/testcases/test4.tig:11.9:syntax error: inserting  SEMICOLON
"
`;

exports[`The textbook tests parses test/testcases/test5.tig correctly 1`] = `""`;

exports[`The textbook tests parses test/testcases/test6.tig correctly 1`] = `
"test/testcases/test6.tig:5.14:syntax error: inserting  PLUS
test/testcases/test6.tig:8.14:syntax error: deleting  LPAREN ID COMMA
test/testcases/test6.tig:8.18:syntax error: deleting  STRING RPAREN
test/testcases/test6.tig:11.13:syntax error: deleting  LPAREN INT COMMA
test/testcases/test6.tig:11.17:syntax error: deleting  STRING RPAREN
"
`;

exports[`The textbook tests parses test/testcases/test7.tig correctly 1`] = `
"test/testcases/test7.tig:5.15:syntax error: inserting  SEMICOLON
test/testcases/test7.tig:8.15:syntax error: deleting  LPAREN ID COMMA
test/testcases/test7.tig:8.19:syntax error: deleting  STRING RPAREN
test/testcases/test7.tig:11.13:syntax error: deleting  LPAREN INT COMMA
test/testcases/test7.tig:11.17:syntax error: deleting  STRING RPAREN
"
`;

exports[`The textbook tests parses test/testcases/test8.tig correctly 1`] = `""`;

exports[`The textbook tests parses test/testcases/test9.tig correctly 1`] = `""`;

exports[`The textbook tests parses test/testcases/test10.tig correctly 1`] = `""`;

exports[`The textbook tests parses test/testcases/test11.tig correctly 1`] = `""`;

exports[`The textbook tests parses test/testcases/test12.tig correctly 1`] = `""`;

exports[`The textbook tests parses test/testcases/test13.tig correctly 1`] = `""`;

exports[`The textbook tests parses test/testcases/test14.tig correctly 1`] = `""`;

exports[`The textbook tests parses test/testcases/test15.tig correctly 1`] = `""`;

exports[`The textbook tests parses test/testcases/test16.tig correctly 1`] = `""`;

exports[`The textbook tests parses test/testcases/test17.tig correctly 1`] = `""`;

exports[`The textbook tests parses test/testcases/test18.tig correctly 1`] = `
"test/testcases/test18.tig:5.15:syntax error: inserting  SEMICOLON
test/testcases/test18.tig:10.15:syntax error: deleting  LPAREN ID COMMA
test/testcases/test18.tig:10.19:syntax error: deleting  STRING RPAREN
test/testcases/test18.tig:13.13:syntax error: deleting  LPAREN INT COMMA
test/testcases/test18.tig:13.17:syntax error: deleting  STRING RPAREN
"
`;

exports[`The textbook tests parses test/testcases/test19.tig correctly 1`] = `
"test/testcases/test19.tig:5.15:syntax error: inserting  SEMICOLON
test/testcases/test19.tig:8.15:syntax error: deleting  LPAREN ID COMMA
test/testcases/test19.tig:8.19:syntax error: deleting  STRING RPAREN
test/testcases/test19.tig:11.13:syntax error: deleting  LPAREN INT COMMA
test/testcases/test19.tig:11.17:syntax error: deleting  STRING RPAREN
"
`;

exports[`The textbook tests parses test/testcases/test20.tig correctly 1`] = `""`;

exports[`The textbook tests parses test/testcases/test21.tig correctly 1`] = `
"test/testcases/test21.tig:8.20:syntax error: inserting  PLUS
test/testcases/test21.tig:11.9:syntax error: inserting  SEMICOLON
"
`;

exports[`The textbook tests parses test/testcases/test22.tig correctly 1`] = `""`;

exports[`The textbook tests parses test/testcases/test23.tig correctly 1`] = `""`;

exports[`The textbook tests parses test/testcases/test26.tig correctly 1`] = `""`;

exports[`The textbook tests parses test/testcases/test27.tig correctly 1`] = `
"test/testcases/test27.tig:7.3:syntax error: inserting  SEMICOLON
"
`;

exports[`The textbook tests parses test/testcases/test28.tig correctly 1`] = `""`;

exports[`The textbook tests parses test/testcases/test29.tig correctly 1`] = `""`;

exports[`The textbook tests parses test/testcases/test31.tig correctly 1`] = `""`;

exports[`The textbook tests parses test/testcases/test32.tig correctly 1`] = `""`;

exports[`The textbook tests parses test/testcases/test33.tig correctly 1`] = `""`;

exports[`The textbook tests parses test/testcases/test34.tig correctly 1`] = `
"test/testcases/test34.tig:5.3:syntax error: deleting  LPAREN STRING COMMA
test/testcases/test34.tig:5.11:syntax error: deleting  STRING RPAREN
"
`;

exports[`The textbook tests parses test/testcases/test35.tig correctly 1`] = `
"test/testcases/test35.tig:5.3:syntax error: inserting  SEMICOLON
"
`;

exports[`The textbook tests parses test/testcases/test36.tig correctly 1`] = `
"test/testcases/test36.tig:5.3:syntax error: deleting  LPAREN INT COMMA
test/testcases/test36.tig:5.6:syntax error: deleting  STRING COMMA INT
test/testcases/test36.tig:5.13:syntax error: deleting  RPAREN
"
`;

exports[`The textbook tests parses test/testcases/test37.tig correctly 1`] = `""`;

exports[`The textbook tests parses test/testcases/test38.tig correctly 1`] = `""`;

exports[`The textbook tests parses test/testcases/test39.tig correctly 1`] = `""`;

exports[`The textbook tests parses test/testcases/test40.tig correctly 1`] = `
"test/testcases/test40.tig:5.3:syntax error: inserting  SEMICOLON
"
`;

exports[`The textbook tests parses test/testcases/test41.tig correctly 1`] = `""`;

exports[`The textbook tests parses test/testcases/test42.tig correctly 1`] = `""`;

exports[`The textbook tests parses test/testcases/test43.tig correctly 1`] = `""`;

exports[`The textbook tests parses test/testcases/test44.tig correctly 1`] = `""`;

exports[`The textbook tests parses test/testcases/test45.tig correctly 1`] = `""`;

exports[`The textbook tests parses test/testcases/test46.tig correctly 1`] = `""`;

exports[`The textbook tests parses test/testcases/test47.tig correctly 1`] = `""`;

exports[`The textbook tests parses test/testcases/test48.tig correctly 1`] = `""`;

exports[`The textbook tests parses test/testcases/test49.tig correctly 1`] = `
"test/testcases/test49.tig:5.18:syntax error: inserting  PLUS
"
`;
